### PR TITLE
Deprecate placeholder nodes

### PR DESCRIPTION
**Pull request by kimdv**

### CODE REVIEW

File: CodeGeneration/Sources/SyntaxSupport/DeclNodes.swift [hunk 1]
```diff
@@ -548,39 +548,6 @@ public let DECL_NODES: [Node] = [
     elementChoices: [.designatedType]
   ),
 
-  Node(
-    kind: .editorPlaceholderDecl,
-    base: .decl,
-    nameForDiagnostics: "editor placeholder",
```
Comment (kimdv) [inline CodeGeneration/Sources/SyntaxSupport/DeclNodes.swift:554]: @ahoppen we don't have nay functionally for setting an complete node as deprecated.
Should we add it?

Comment (ahoppen) [reply CodeGeneration/Sources/SyntaxSupport/DeclNodes.swift:554]: Sorry for the delayed response. We needed to discuss how to handle this. 

We should deprecate the nodes. The parser should just stop generating them but that way they will still exist for anyone using swift-syntax for code generation and it will make transitioning to a new swift-syntax version easier because the deprecation warnings can guide you in the right direction.

Comment (kimdv) [reply CodeGeneration/Sources/SyntaxSupport/DeclNodes.swift:554]: Would the best approach to to just copy the needed code and place it in the deprecated files or should we introduce a new field on the `Node` where we can deprecate it?

Because we also need to ensure the parser have the `open func visit(_ node: EditorPlaceholderDeclSyntax) -> DeclSyntax` that is deprecated. 
Otherwise it will break

Comment (ahoppen) [reply CodeGeneration/Sources/SyntaxSupport/DeclNodes.swift:554]: I think we need to properly generate all the generated cases/functions using CodeGeneration. Maintaining them manually won’t scale, I think.

Comment (kimdv) [reply CodeGeneration/Sources/SyntaxSupport/DeclNodes.swift:554]: As I tought. Just needed to confirm

**Comment by kimdv** [comment]
It will introduce a lot of warnings inside the project. 
Not sure if we should try to reduce some of them?

File: CodeGeneration/Sources/SyntaxSupport/Node.swift [hunk 2]
```diff
@@ -106,10 +109,16 @@ public class Node {
...
         experimentalSPI.with(\.trailingTrivia, .newline)
       }
+
       if forRaw {
         "@_spi(RawSyntax)"
       }
+
+      if isDeprecated, !ignoreAvailableAttribute {
+        #"@available(*, deprecated, message: "Will be removed in a future version.")"#
+      }
```
Comment (kimdv) [inline CodeGeneration/Sources/SyntaxSupport/Node.swift:119]: A single place there is a enum that conforms to `CaseItaretable` and then we cannot apply this attribute.

File: Tests/SwiftParserTest/DeclarationTests.swift [hunk 3]
```diff
@@ -2272,7 +2272,7 @@ final class DeclarationTests: ParserTestCase {
         1️⃣<#code#>
       }
       """,
-      substructure: MemberBlockItemSyntax(decl: EditorPlaceholderDeclSyntax(placeholder: .identifier("<#code#>"))),
+      // substructure: MemberBlockItemSyntax(decl: DeclReferenceExprSyntax(baseName: .identifier("<#code#>"))),
```
Comment (ahoppen) [inline Tests/SwiftParserTest/DeclarationTests.swift:2275]: Should this really be a comment?

File: Tests/SwiftParserTest/translated/TrailingClosuresTests.swift [hunk 4]
```diff
@@ -185,14 +185,14 @@ final class TrailingClosuresTests: ParserTestCase {
...
-              item: .init(EditorPlaceholderExprSyntax(placeholder: .identifier("<#T##() -> Void#>")))
+              item: .init(DeclReferenceExprSyntax(baseName: .identifier("<#T##() -> Void#>")))
             )
           ]),
           rightBrace: .rightBraceToken(presence: .missing)
         )
       ),
       diagnostics: [
-        DiagnosticSpec(message: "editor placeholder in source file")
+        DiagnosticSpec(locationMarker: "1️⃣", message: "editor placeholder in source file")
```
Comment (ahoppen) [inline Tests/SwiftParserTest/translated/TrailingClosuresTests.swift:195]: Is this change needed?

**Comment by ahoppen** [comment]
> It will introduce a lot of warnings inside the project.
> Not sure if we should try to reduce some of them?

Sorry, only read this comment now by chance.

swift-syntax should definitely build without warnings after this change. It’s OK for dependent projects to have deprecation warnings but swift-syntax should be warning-free.

**Comment by kimdv** [comment]
Okay @ahoppen I've now updated the parser, so it don't parse the placeholder anymore. 
As we talked about, they will not be marked as deprecated because they will make warnings in the project

**Comment by ahoppen** [comment]
As we discussed offline, could you add a warning doc comment to the editor placeholder nodes? Something along the lines of:
```swift
/// - Warning: Theses `EditorPlaceholderDeclSyntax` node is not generated by the parser anymore. Placeholders are represented by a `MissingDeclSyntax`.
```

**Comment by kimdv** [comment]
> As we discussed offline, could you add a warning doc comment to the editor placeholder nodes? Something along the lines of:
> 
> ```swift
> /// - Warning: Theses `EditorPlaceholderDeclSyntax` node is not generated by the parser anymore. Placeholders are represented by a `MissingDeclSyntax`.
> ```

Something like this?

File: Release Notes/511.md [hunk 5]
```diff
@@ -30,6 +30,10 @@
   - Issue: https://github.com/apple/swift-syntax/issues/2267
   - Pull request: https://github.com/apple/swift-syntax/pull/2272
 
+- `EditorPlaceholderExprSyntax` & `EditorPlaceholderExprSyntax`:
+  - Description: `EditorPlaceholderExprSyntax` & `EditorPlaceholderExprSyntax` are now deprecated and not parsed anymore. They will now be parsed as identifiers.
```
Comment (bnbarham) [inline Release Notes/511.md:34]: `Expr` twice, presumably wanted `Decl` as well? Super nitpick, but I'd also just use `and` rather than `&`.

```suggestion
- `EditorPlaceholderDeclSyntax` and `EditorPlaceholderExprSyntax` deprecated:
  - Description: `EditorPlaceholderExprSyntax` and `EditorPlaceholderExprSyntax` are now deprecated and placeholders are instead parsed as identifiers within a `MissingDeclSyntax` or `DeclReferenceExprSyntax`.
```

Comment (kimdv) [reply Release Notes/511.md]: I tell you. This time I didn't want any copy/paste error. But it happened again.. 🤦‍♂️

but it's fixed and nothing is to nitpick

**Review by ahoppen — APPROVED** [review]

**Comment by kimdv** [comment]
@swift-ci please test

**Comment by kimdv** [comment]
@swift-ci please test Windows
